File: src/hid/devices/as5600.ts
import type { VirtualI2cDevice } from "../i2c-device.ts";

/**
 * A synthetic AS5600 magnetic encoder.
 *
 * A register file with a shaft behind it, and -- more usefully -- a magnet that
 * can be put at the wrong distance. The angle registers hand back a plausible
 * number in every case, including with no magnet at all, which is exactly the
 * failure the panel exists to catch, so the emulator has to be able to produce
 * it: `gap` drives the automatic gain and the status bits the same way moving a
 * real magnet would.
 */

const REG_ZPOS_H = 0x01;
const REG_STATUS = 0x0b;
const REG_RAWANGLE_H = 0x0c;
const REG_ANGLE_H = 0x0e;
const REG_AGC = 0x1a;
const REG_MAGNITUDE_H = 0x1b;
const REGISTER_COUNT = 0x100;

const STATUS_MH = 1 << 3; // gain bottomed out: magnet too strong or too close
const STATUS_ML = 1 << 4; // gain maxed out: too weak or too far
const STATUS_MD = 1 << 5; // a magnet is there at all

const FULL_SCALE = 4096;

export interface As5600Options {
  address?: number;
  /** Shaft angle in degrees. Wraps freely; turn counting is the driver's job. */
  degrees?: number;
  /**
   * How far the magnet sits from the die, 0 (touching) to 1 (gone). The middle
   * of the range is a good gap; the ends saturate the gain and raise ML or MH.
   */
  gap?: number;
  /** No magnet at all. The angle registers still return something. */
  magnetAbsent?: boolean;
}

export class VirtualAs5600 implements VirtualI2cDevice {
  readonly address: number;
  readonly name = "AS5600 magnetic encoder";

  readonly #registers = new Uint8Array(REGISTER_COUNT);
  degrees: number;
  gap: number;
  magnetAbsent: boolean;
  #pointer = 0;

  constructor(options: As5600Options = {}) {
    this.address = options.address ?? 0x36;
    this.degrees = options.degrees ?? 0;
    this.gap = options.gap ?? 0.5;
    this.magnetAbsent = options.magnetAbsent ?? false;
  }

  /** Turn the shaft, in degrees. Accepts any value and wraps. */
  rotate(by: number): void {
    this.degrees = (((this.degrees + by) % 360) + 360) % 360;
  }

  write(data: Uint8Array): void {
    const register = data[0];
    if (register === undefined) return; // address probe
    const payload = data.subarray(1);
    if (payload.length === 0) {
      this.#pointer = register;
      return;
    }
    for (let i = 0; i < payload.length; i++) {
      this.#registers[(register + i) % REGISTER_COUNT] = payload[i] ?? 0;
    }
    this.#pointer = (register + payload.length) % REGISTER_COUNT;
  }

  read(length: number): Uint8Array {
    const out = new Uint8Array(length);
    for (let i = 0; i < length; i++) {
      out[i] = this.#value((this.#pointer + i) % REGISTER_COUNT);
    }
    this.#pointer = (this.#pointer + length) % REGISTER_COUNT;
    return out;
  }

  #value(at: number): number {
    const raw = Math.round((this.degrees / 360) * FULL_SCALE) % FULL_SCALE;
    if (at === REG_RAWANGLE_H) return (raw >> 8) & 0x0f;
    if (at === REG_RAWANGLE_H + 1) return raw & 0xff;
    if (at === REG_ANGLE_H || at === REG_ANGLE_H + 1) {
      // The scaled angle is the raw one shifted by ZPOS, which is the whole
      // point of having both -- one is the shaft, the other is your datum.
      const zero = ((this.#registers[REG_ZPOS_H] ?? 0) << 8) | (this.#registers[REG_ZPOS_H + 1] ?? 0);
      const scaled = (raw - (zero & 0x0fff) + FULL_SCALE) % FULL_SCALE;
      return at === REG_ANGLE_H ? (scaled >> 8) & 0x0f : scaled & 0xff;
    }
    if (at === REG_STATUS) return this.#status();
    if (at === REG_AGC) return this.#agc();
    if (at === REG_MAGNITUDE_H) return (this.#magnitude() >> 8) & 0x0f;
    if (at === REG_MAGNITUDE_H + 1) return this.#magnitude() & 0xff;
    return this.#registers[at] ?? 0;
  }

  /**
   * Automatic gain. The chip winds it up as the field weakens, so it rises
   * with the gap -- which is what makes it a proxy for magnet distance.
   */
  #agc(): number {
    if (this.magnetAbsent) return 255;
    return Math.max(0, Math.min(255, Math.round(this.gap * 255)));
  }

  #magnitude(): number {
    if (this.magnetAbsent) return 0;
    return Math.round((1 - this.gap) * FULL_SCALE);
  }

  #status(): number {
    if (this.magnetAbsent) return STATUS_ML; // gain maxed, nothing found
    let status = STATUS_MD;
    if (this.#agc() >= 255) status |= STATUS_ML;
    if (this.#agc() <= 0) status |= STATUS_MH;
    return status;
  }
}

File: python/webblinka/drivers/as5600.py
"""AS5600 magnetic rotary position sensor, via the stock adafruit_as5600 library.

**An angle reading always looks fine.** The part will hand back a perfectly
plausible number between 0 and 360 degrees whether the magnet is at the right
distance, too far, too close, off-centre, or absent entirely -- in the last case
it is reporting noise, with no hint in the value itself. So the panel leads with
the magnet's health, and the driver reports the three status bits alongside the
automatic gain, which is the actual measure of how far away the magnet is: the
chip winds the gain up as the field weakens, so a figure pinned at either end
means the gap is wrong in one direction or the other.

**Turn counting is the driver's job, not the chip's.** The AS5600 wraps at 360
and has no idea how many times it has been round. Accumulating that is easy to
get wrong -- the give-away is a jump of more than half a revolution between two
readings, which is a wrap rather than a genuine movement -- and it is the thing
you actually want from a shaft encoder.
"""

from __future__ import annotations

import time
from typing import Any

from .base import Driver, register

DEFAULT_ADDRESS = 0x36

#: The angle registers are 12 bits, so a full turn is this many counts.
FULL_SCALE = 4096

#: A step larger than half a revolution between consecutive readings is a wrap,
#: not a movement. That holds as long as the shaft turns less than half a turn
#: between polls -- at the panel's rate, under about 150 rpm.
WRAP_THRESHOLD = FULL_SCALE // 2


@register("as5600")
class As5600(Driver):
    """ams AS5600 12-bit magnetic encoder."""

    def __init__(self, bus, address: int = DEFAULT_ADDRESS) -> None:
        super().__init__(bus, address)
        self._sensor = None
        self._turns = 0
        self._last_raw: int | None = None
        self._last_at: float | None = None
        self._speed_dps = 0.0

    def start(self) -> dict[str, Any]:
        import adafruit_as5600

        self._sensor = adafruit_as5600.AS5600(self.bus, address=self.address)
        self._reset_tracking()
        return {"address": self.address, "fullScale": FULL_SCALE}

    def stop(self) -> None:
        self._sensor = None

    def command(self, name: str, args: list[Any]) -> Any:
        sensor = self._require()
        if name == "set_zero_here":
            # ZPOS shifts the scaled angle so that wherever the shaft is now
            # reads zero. The raw angle is untouched, which is why the panel
            # shows both -- one is the shaft, the other is your choice of datum.
            sensor.z_position = sensor.raw_angle
            return self.poll()
        if name == "reset_turns":
            self._reset_tracking()
            return self.poll()
        return super().command(name, args)

    def poll(self) -> dict[str, Any]:
        sensor = self._require()
        now = time.monotonic()

        raw = sensor.raw_angle
        scaled = sensor.angle
        agc = sensor.agc

        # Accumulate turns from the *raw* angle: the scaled one moves when the
        # zero position is changed, which would look like a giant jump.
        if self._last_raw is not None:
            delta = raw - self._last_raw
            if delta > WRAP_THRESHOLD:
                self._turns -= 1
                delta -= FULL_SCALE
            elif delta < -WRAP_THRESHOLD:
                self._turns += 1
                delta += FULL_SCALE
            elapsed = now - (self._last_at or now)
            if elapsed > 0:
                degrees = delta * 360 / FULL_SCALE
                # Smoothed, because one poll's worth of quantisation over a
                # short interval is a wild speed. Two counts at 10 Hz is 17
                # degrees per second of pure noise.
                self._speed_dps = 0.7 * self._speed_dps + 0.3 * (degrees / elapsed)
        self._last_raw = raw
        self._last_at = now

        detected = bool(sensor.magnet_detected)
        too_weak = bool(sensor.max_gain_overflow)
        too_strong = bool(sensor.min_gain_overflow)

        return {
            "raw": raw,
            "scaled": scaled,
            "rawDegrees": raw * 360 / FULL_SCALE,
            "degrees": scaled * 360 / FULL_SCALE,
            "turns": self._turns,
            "continuousDegrees": self._turns * 360 + scaled * 360 / FULL_SCALE,
            "speedDps": self._speed_dps,
            "speedRpm": self._speed_dps / 6,
            "zeroPosition": sensor.z_position,
            "agc": agc,
            "magnitude": sensor.magnitude,
            "magnetDetected": detected,
            "magnetTooWeak": too_weak,
            "magnetTooStrong": too_strong,
            "magnet": _magnet_verdict(detected, too_weak, too_strong, agc),
        }

    def _reset_tracking(self) -> None:
        self._turns = 0
        self._last_raw = None
        self._last_at = None
        self._speed_dps = 0.0

    def _require(self):
        if self._sensor is None:
            raise RuntimeError("AS5600 not started")
        return self._sensor


def _magnet_verdict(
    detected: bool, too_weak: bool, too_strong: bool, agc: int
) -> dict[str, Any]:
    """What the status bits and the gain say about the magnet.

    The gain is the informative one. The chip raises it to compensate for a
    weaker field, so a high figure means the magnet is too far away and a low
    one means it is too close -- and either extreme means the gain has run out
    of room, which is what the ML and MH bits are reporting.
    """
    if not detected:
        return {
            "state": "absent",
            "tone": "error",
            "text": "No magnet detected — the angle below is noise, not a position.",
        }
    if too_weak:
        return {
            "state": "weak",
            "tone": "warn",
            "text": (
                f"Magnet too weak or too far (gain {agc} is at its ceiling). "
                "Move it closer, or use a stronger one."
            ),
        }
    if too_strong:
        return {
            "state": "strong",
            "tone": "warn",
            "text": (
                f"Magnet too strong or too close (gain {agc} has bottomed out). "
                "Move it further away."
            ),
        }
    return {
        "state": "ok",
        "tone": "ok",
        "text": f"Magnet in range, gain {agc}. Mid-range gain means a good gap.",
    }
